File: src/main/java/de/aquafun3d/vanilla/commands/WaypointCommand.java
package de.aquafun3d.vanilla.commands;

import de.aquafun3d.vanilla.Main;
import de.aquafun3d.vanilla.utils.Utils;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.ArrayList;

public class WaypointCommand implements CommandExecutor {

	private int clear = 0;
	private int timeClear = 0;

	@Override
	public boolean onCommand(CommandSender sender, Command command, String label, String[] args){
		if(sender instanceof Player){
			Player player = (Player) sender;
			if(args.length == 0){
				Utils.send(player, ChatColor.RED + "/waypoint <name> or list/remove/clear");
				return false;
			}
			switch(args[0]){
				case "list" -> {
					ArrayList<String> list = Main.vanillaConfig.getAll(".waypoints");
					Utils.send(player,ChatColor.GREEN + "Available Waypoints are:");
					if(list == null || list.size() == 0){
						Utils.send(player,ChatColor.RED + "NONE");
					}else{
						StringBuilder out = new StringBuilder();
						for(int i = 0; i < list.size();i++){
							if ( i != list.size() - 1) {
								out.append(list.get(i)).append(", ");
							}else{
								out.append(list.get(i));
							}
						}
						Utils.send(player,ChatColor.LIGHT_PURPLE.toString() + out);
					}

				}
				case "remove" -> {
					if(player.isOp()){
						String s = null;
						if(args.length == 3){
							s = args[1] + "_" + args[2];
						}else if(args.length == 2){
							s = args[1];
						}else{
							Utils.send(player, ChatColor.YELLOW + "Use /wp remove <name>");
						}
						try {
							Main.vanillaConfig.set("waypoints." + s, null);
							} catch (IOException e) {
							e.printStackTrace();
						}
						Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GOLD + " deleted");
					}else {
						Utils.send(player,ChatColor.GOLD + "You don't have Op permissions");
					}
				}
				case "clear" -> {
					if (player.isOp()) {
						if(LocalDateTime.now().getMinute() - timeClear >= 1){
							clear = 0;
						}
						if(clear == 0){
							Utils.send(player,ChatColor.RED + "Are you Sure? " + ChatColor.GREEN + "Run Command again.");
							clear = 1;
							timeClear = LocalDateTime.now().getMinute();
						}else if(clear == 1) {
							try {
								Main.vanillaConfig.set("waypoints", null);
							} catch (IOException e) {
								e.printStackTrace();
							}
							Utils.send(player,ChatColor.DARK_RED + "All waypoints deleted!");
							clear = 0;
						}
					} else {
						Utils.send(player,ChatColor.GOLD + "You are not Admin!");
					}
				}
				default -> {
					String s;
					if (args.length >= 2) {
						s = args[0] + "_" + args[1];
					} else {
						s = args[0];
					}
					if(Main.vanillaConfig.contains("waypoints." + s)){
						String w = Main.vanillaConfig.getString("waypoints." + s + ".world");
						int x = Main.vanillaConfig.getInt("waypoints." + s + ".x");
						int y = Main.vanillaConfig.getInt("waypoints." + s + ".y");
						int z = Main.vanillaConfig.getInt("waypoints." + s + ".z");
						switch(w){
							case "world" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in Overworld");
							case "world_nether" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in End");
							case "world_the_end" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in Nether");
						}
					}else{
						int x = player.getLocation().getBlockX();
						int y = player.getLocation().getBlockY();
						int z = player.getLocation().getBlockZ();
						try {
							Main.vanillaConfig.set("waypoints."+ s + ".world", player.getWorld().getName());
							Main.vanillaConfig.set("waypoints."+ s + ".x", x);
							Main.vanillaConfig.set("waypoints."+ s + ".y", y);
							Main.vanillaConfig.set("waypoints."+ s + ".z", z);
						} catch (IOException e) {
							e.printStackTrace();
						}
						switch(player.getWorld().getName()){
							case "world" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in Overworld by " + ChatColor.AQUA + player.getName());
							case "world_nether" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in Nether by " + ChatColor.AQUA + player.getName());
							case "world_the_end" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + s + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in End by " + ChatColor.AQUA + player.getName());
						}
					}
				}
			}
		}
		return false;
	}
}

File: src/main/kotlin/de/aquafun3d/vanilla/commands/WaypointCommand.kt
package de.aquafun3d.vanilla.commands

import de.aquafun3d.vanilla.Main
import de.aquafun3d.vanilla.utils.Utils
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.time.LocalDateTime

class WaypointCommand :CommandExecutor{

	private var clear: Int = 0
	private var timeClear: Int = 0

	//handles waypoints
	override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>): Boolean {
		if (sender is Player) {
			val player: Player = sender
			if (args.isEmpty()) {
				player.sendMessage(Utils.PREFIX + ChatColor.RED + "/waypoint <name> or list/remove/clear")
			}
			when(args[0]){
				"list" -> {
					val list: ArrayList<String>? = Main.vanillaConfig?.getAll("waypoints.")
					player.sendMessage(Utils.PREFIX + ChatColor.GREEN  + "Available Waypoints are:")
					if(list == null || list.size == 0) {
						player.sendMessage(Utils.PREFIX + ChatColor.RED + "NONE")
					}else {
						var out = ""
						for (i in 0 until list.size) {
							out += if ( i != list.size - 1) {
								list[i] + ", "
							}else{
								list[i]
							}
						}
						player.sendMessage(Utils.PREFIX + ChatColor.LIGHT_PURPLE + out)
					}
				}
				"remove" -> {
					if (player.isOp){
						val str: String = if (args.size == 3){
							args[1] + "_" + args[2]
						}else{
							args[1]
						}
						Main.vanillaConfig?.set("waypoints.$str",null)
						Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GOLD + " deleted")
					}else {
						player.sendMessage(Utils.PREFIX + ChatColor.GOLD + "You don't have Op permissions")
					}
				}
				"clear" -> {
					if (player.isOp) {
						if(LocalDateTime.now().minute - timeClear >= 1){
							clear = 0
						}
						if(clear == 0){
							player.sendMessage(Utils.PREFIX + ChatColor.RED + "Are you Sure? " + ChatColor.GREEN + "Run Command again.")
							clear = 1
							timeClear = LocalDateTime.now().minute
						}else if(clear == 1) {
							Main.vanillaConfig?.set("waypoints", null)
							player.sendMessage(Utils.PREFIX + ChatColor.DARK_RED + "All waypoints deleted!")
							clear = 0
						}
					} else {
						player.sendMessage(Utils.PREFIX + ChatColor.GOLD + "You are not Admin!")
					}
				}
				else -> {
					val str = if (args.size >= 2) {
						args[0] + "_" + args[1]
					} else {
						args[0]
					}
					if(Main.vanillaConfig?.contains("waypoints.$str") == true){
						val wp: String = Main.vanillaConfig?.getString("waypoints.$str.world")!!
						val x: Int = Main.vanillaConfig?.getDouble("waypoints.$str.x")!!.toInt()
						val y: Int = Main.vanillaConfig?.getDouble("waypoints.$str.y")!!.toInt()
						val z: Int = Main.vanillaConfig?.getDouble("waypoints.$str.z")!!.toInt()
						when (wp){
							"world" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in overworld")
							"world_nether" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in nether")
							"world_the_end" -> 	Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in the end")
						}
					}else {
						Main.vanillaConfig?.set("waypoints.$str.world", player.world.name)
						Main.vanillaConfig?.set("waypoints.$str.x", player.location.x)
						Main.vanillaConfig?.set("waypoints.$str.y", player.location.y)
						Main.vanillaConfig?.set("waypoints.$str.z", player.location.z)
						val x = player.location.x.toInt()
						val y = player.location.y.toInt()
						val z = player.location.z.toInt()
						when (player.world.name) {
							"world" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in overworld by " + ChatColor.AQUA + player.name)
							"world_nether" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in nether by " + ChatColor.AQUA + player.name)
							"world_the_end" -> Utils.atAll("" + ChatColor.GOLD + "Waypoint " + ChatColor.LIGHT_PURPLE + str + ChatColor.GREEN + " " + x + ", " + y + ", " + z + ChatColor.GOLD + " in the end by " + ChatColor.AQUA + player.name)
						}
					}
				}
			}

		}
		return false
	}
}